File: parser.cpp
#include "common.h"
#include "string.h"
#include "parser.h"

extern const char *print_keyword;

void Parser::init(const char *source) {
    lex.init(source);
    lex.eat_token();
}

Expr *Parser::alloc_expr(Expr expr) {
    Expr *buf = (Expr *) xmalloc(sizeof(Expr));
    memcpy(buf, &expr, sizeof(Expr));
    return buf;
}

Stmt *Parser::alloc_stmt(Stmt stmt) {
    Stmt *buf = (Stmt *) xmalloc(sizeof(Stmt));
    memcpy(buf, &stmt, sizeof(Stmt));
    return buf;
}

Expr *Parser::parse_expr() {
    return parse_expr_equality();
}

Expr *Parser::parse_expr_equality() {
    Expr *expr = parse_expr_comparison();

    while (lex.match(EQUAL_EQUAL) || lex.match(BANG_EQUAL)) {
        Token_Type op = lex.prev_token.type;
        Expr *right = parse_expr_comparison();
        expr = alloc_expr(Expr{BINARY_EXPR, {op, expr, right}});
    }

    return expr;
}

Expr *Parser::parse_expr_comparison() {
    Expr *expr = parse_expr_term();

    while (lex.match((Token_Type)'>') || lex.match((Token_Type)'<') || 
           lex.match(GREATER_EQUAL) || lex.match(LESS_EQUAL)) {
        Token_Type op = lex.prev_token.type;
        Expr *right = parse_expr_term();
        expr = alloc_expr(Expr{BINARY_EXPR, {op, expr, right}});
    }

    return expr;
}

Expr *Parser::parse_expr_term() {
    Expr *expr = parse_expr_factor();

    while (lex.match((Token_Type)'+') || lex.match((Token_Type)'-')) {
        Token_Type op = lex.prev_token.type;
        Expr *right = parse_expr_factor();
        expr = alloc_expr(Expr{BINARY_EXPR, {op, expr, right}});
    }

    return expr;
}

Expr *Parser::parse_expr_factor() {
    Expr *expr = parse_expr_unary();

    while (lex.match((Token_Type)'*') || lex.match((Token_Type)'/')) {
        Token_Type op = lex.prev_token.type;
        Expr *right = parse_expr_unary();
        expr = alloc_expr(Expr{BINARY_EXPR, {op, expr, right}});
    }

    return expr;
}

Expr *Parser::parse_expr_unary() {
    if (lex.match((Token_Type)'!') || lex.match((Token_Type)'-')) {
        Token_Type op = lex.prev_token.type;
        Expr *right = parse_expr_unary();
        return alloc_expr(Expr{UNARY_EXPR, {op, right}});
    }

    return parse_expr_primary();
}

Expr *Parser::parse_expr_primary() {
    if (lex.match(FALSE)) return alloc_expr(Expr{LITERAL_EXPR, .number = false});
    if (lex.match(TRUE))  return alloc_expr(Expr{LITERAL_EXPR, .number = true});
    if (lex.match(NIL))   return alloc_expr(Expr{LITERAL_EXPR, .number = 0});
    if (lex.match(NUMBER))
        return alloc_expr(Expr{LITERAL_EXPR, .number = lex.token.val.int_val});
    if (lex.match(STRING))
        return alloc_expr(Expr{LITERAL_EXPR, .name = lex.token.val.name});

    if (lex.match((Token_Type)'(')) {
        Expr *expr = parse_expr();
        if (!lex.match((Token_Type)')')) {
            printf("Expect ')' after expression.");
            exit(1);
        }
        lex.eat_token();
        return alloc_expr(Expr{GROUPING_EXPR, .group = expr});
    }

    printf("[!] You fucked up somewhere\n");
    exit(1);
}

Stmt *Parser::parse_stmt_print() {
    Expr *val = parse_expr();
    if (!lex.match((Token_Type)';')) {
        fprintf(stderr, "Expected ';' after value.\n");
        return alloc_stmt(Stmt{STMT_NONE});
    }
    return alloc_stmt(Stmt{STMT_PRINT, val});
}

Stmt *Parser::parse_stmt_expr() {
    Expr *expr = parse_expr();
    if (!lex.match((Token_Type)';')) {
        fprintf(stderr, "Expected ';' after expression.\n");
        return alloc_stmt(Stmt{STMT_NONE});
    }
    return alloc_stmt(Stmt{STMT_EXPR, expr});
}

Stmt *Parser::parse_stmt() {
    Stmt *stmt = NULL;
    if (lex.is_token((Token_Type)'\0')) {
        stmt = NULL;
    } else if (lex.match_keyword(print_keyword)) {
        stmt = parse_stmt_print();
    } else {
        stmt = parse_stmt_expr();
    }

    return stmt;
}

Stmt *Parser::parse() {
    return parse_stmt();
}

void Parser::print_ast(Expr *root) {
    switch (root->type) {
    case BINARY_EXPR:
        printf("(%c ", root->bin.op);
        print_ast(root->bin.left);
        printf(" ");
        print_ast(root->bin.right);
        printf(") ");
        break;
    case LITERAL_EXPR:
        printf("%d", root->number);
        break;
    }
}

File: parser.go
package main

import "errors"

var ParseError = errors.New("error while parsing")

type Parser struct {
	tokens  []Token
	current int
}

func (p *Parser) primary() Expr {
	if p.match(False) {
		return LiteralExpr{value: false}
	}
	if p.match(True) {
		return LiteralExpr{value: true}
	}
	if p.match(Nil) {
		return LiteralExpr{value: nil}
	}

	if p.match(Number, String) {
		return LiteralExpr{value: p.previous().literal}
	}

	if p.match(LeftParen) {
		expr := p.expression()
		p.consume(RightParen, "Expect ')' after expression")
		return GroupingExpr{expr: expr}
	}
	return nil
}

func (p *Parser) unary() Expr {
	if p.match(Bang, Minus) {
		op := p.previous()
		right := p.unary()
		return UnaryExpr{
			operator: op,
			right:    right,
		}
	}
	return p.primary()
}

func (p *Parser) factor() Expr {
	expr := p.unary()

	for p.match(Star, Slash) {
		op := p.previous()
		right := p.unary()
		expr = BinaryExpr{
			left:     expr,
			operator: op,
			right:    right,
		}
	}

	return expr
}

func (p *Parser) term() Expr {
	expr := p.factor()

	for p.match(Minus, Plus) {
		op := p.previous()
		right := p.factor()
		expr = BinaryExpr{
			left:     expr,
			operator: op,
			right:    right,
		}
	}

	return expr
}

func (p *Parser) comparison() Expr {
	expr := p.term()

	for p.match(Less, LessEqual, Greater, GreaterEqual) {
		op := p.previous()
		right := p.term()
		expr = BinaryExpr{
			left:     expr,
			operator: op,
			right:    right,
		}
	}

	return expr
}

func (p *Parser) equality() Expr {
	expr := p.comparison()

	for p.match(EqualEqual, BangEqual) {
		op := p.previous()
		right := p.comparison()
		expr = BinaryExpr{
			left:     expr,
			operator: op,
			right:    right,
		}
	}

	return expr
}

func (p *Parser) expression() Expr {
	return p.equality()
}

func (p *Parser) consume(t TokenType, errorMessage string) Token {
	if p.check(t) {
		return p.advance()
	}

	lumi.error(p.peek(), errorMessage)
	panic(ParseError)
}

func (p *Parser) match(tt ...TokenType) bool {
	for _, t := range tt {
		if p.check(t) {
			p.advance()
			return true
		}
	}
	return false
}

func (p *Parser) advance() Token {
	if !p.isAtEnd() {
		p.current++
	}
	return p.previous()
}

func (p *Parser) check(t TokenType) bool {
	if p.isAtEnd() {
		return false
	}
	return p.peek().tokenType == t
}

func (p *Parser) previous() Token {
	return p.tokens[p.current-1]
}

func (p *Parser) peek() Token {
	return p.tokens[p.current]
}

func (p *Parser) isAtEnd() bool {
	return p.peek().tokenType == Eof
}
